rids/rules.py

```diff
@@ -0,0 +1,111 @@
+# Copyright 2022 Jigsaw Operations LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Generalizes the various formats of IOCs into Rule and RuleSet.
+
+Rule is a value type that represents evaluation on a single rule.  RuleSet
+instances represent a set of independent and overlapping Rule instances.
+
+RuleSet should typically be constructed from one of the flavors of parsers
+in `rids.iocs` or merged from combining a pair of RuleSet instances.  Its
+
+Evaluation of the RuleSet is done with the ProcessHandshake(...) and
+ProcessEndpoint(...) methods which take an observation and return a list
+of zero or more Event instances with their matching rules in the RuleSet.
+"""
+
+
+import collections
+from dataclasses import dataclass
+
+from rids import observations
+
+
+class RuleSet:
```

Comment:
Can you separate the Ip rules from the suspicious tls connections detection? They can go in separate files.

The conflation makes one think that one affects the other somehow, or that they share state. I don't see the value of keeping them together. It's easier to think about them separately.

I'd go further and call it `IpRuleSet`, so it doesn't ground our thinking when thinking about other IOCs.

The suspicious TLS connections is a single rule, so it doesn't have to be a RuleSet.

Reply:
I have it this way so that fetching from all IOCs doesn't have to know what kinds of rules are included in each feed (some feeds contain more than one type of rule), and only the specific parsers have to be concerned with how the RuleSet gets updated.  What you propose would leak a lot of abstraction into the IOC fetcher/manager and involve multiple RuleSets being passed to it as arguments.  There would be separation of the ruleset types but it would add clutter to the RuleSet construction and use.

Suspicious TLS connections should expand to look at JA3/JA3S fingerprints too, so it's not really just one rule.  Or is that another ruleset type?  There's also domain name IOCs I've already started working on, which would have to be another rule type.  And suricata rules may add to more than one of these.  There are also IOCs for subnets, which could probably be folded into the IpRuleSet but the checking logic would be a little different.  If it weren't for suricata/snort rules I might agree with having the config break them out to different types so that we could more easily manage the different rulesets, but I think this organization with a container object and methods for checking against different rule types is better.

I agree that we can consolidate the IpPacket / IpPacketMonitor / IpRule into one file, likewise for the TLS-related monitoring classes.  I've tried to strike a balance between the inherent n:n mapping of feed formats to IOC types and the benefits of a single ruleset to manage.  I did get to dismantle the rules.py and observations.py modules, so now the modules are either a kind of IOC feed format to process or a kind of monitoring.